<?php
namespace {|{name}|};

use Exception;
use shared\Service;

include_once "{|{Name}|}Repository.php";

class {|{Name}|}Service extends Service
{
    public function __construct()
    {
        parent::__construct(new {|{Name}|}ModelType());
    }

    /**
     * @throws Exception
     */
    public function getAll(): array
    {
        $repo = new {|{Name}|}Repository();

        ${|{name}|} = [];
        $result = $repo->readAll("unable to find any {|{name}|}");

        foreach($result as $row) {
            ${|{name}|}[] = $row;
        }

        return ${|{name}|};
    }

    /**
     * @throws Exception
     */
    public function findById(int $id): array
    {
        $repo = new {|{Name}|}Repository();
        return $repo->read($id, "{|{name}|} not found");
    }

    /**
     * @throws Exception
     */
    public function save(object $input): void
    {
        $repo = new {|{Name}|}Repository();
        $toquery = $this->modelType->isValidType($input);
        $repo->create($toquery, "unable to create {|{name}|}");
    }

    /**
     * @throws Exception
     */
    public function update(object $input): void
    {
        $repo = new {|{Name}|}Repository();
        $toquery = $this->modelType->isValidType($input);
        $repo->update($toquery, "unable to update {|{name}|}");
    }

    /**
     * @throws Exception
     */
    public function delete(int $id): void
    {
        $repo = new {|{Name}|}Repository();
        $repo->delete($id);
    }
}